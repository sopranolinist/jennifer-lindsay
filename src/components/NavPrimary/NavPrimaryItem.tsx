import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import React, { ReactElement } from 'react';
import { RouteComponentProps, withRouter } from 'react-router-dom';
import { isSelected } from '../../helpers/NavHelper';

interface Props extends RouteComponentProps {
  label: string;
  icon?: () => ReactElement;
  pathname?: string;
  path?: string;
  subnav?: () => void;
  external?: boolean;
  name: string;
  children?: ReactElement;
  closeDrawer: () => void;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      paddingTop: 16,
      paddingBottom: 16,
    },
  }),
);

const NavPrimaryItem = ({
  label,
  icon,
  pathname,
  path,
  subnav,
  external,
  name,
  history,
  location,
  children,
  closeDrawer,
}: Props) => {
  const classes = useStyles();

  const handleClick = (event: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => {
    if (external) {
      closeDrawer();
    } else if (subnav) {
      subnav();
    } else {
      event.preventDefault();
      closeDrawer();
      history.push({
        pathname,
      });
    }
  };
  return (
    <>
      <ListItem
        button
        data-testid={`navprimaryitem-${name}`}
        selected={isSelected(path, location.pathname)}
        component="a"
        href={pathname}
        onClick={handleClick}
        className={classes.root}
        {...(external
          ? {
              target: '_blank',
            }
          : '')}
      >
        {icon && <ListItemIcon>{icon()}</ListItemIcon>}
        <ListItemText primary={label} inset={!icon} />
      </ListItem>
      {children}
    </>
  );
};

export default withRouter(NavPrimaryItem);
