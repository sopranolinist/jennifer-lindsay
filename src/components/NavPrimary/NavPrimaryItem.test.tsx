import { fireEvent, render, waitForElement } from '@testing-library/react';
import 'jest-dom/extend-expect';
import React from 'react';
import { MemoryRouter as Router, Route } from 'react-router-dom';
import {
  authExamples,
  getFullyWrappedComponent
} from '../../helpers/TestHelper';
import NavPrimaryItem from './NavPrimaryItem';

const TestOne = () => <span>test 1</span>;
const TestTwo = () => <span>test 2</span>;
const Frag = () => <React.Fragment />;

describe('Screens - NavPrimaryItem', () => {
  const { authContextProps, authObjProps } = authExamples.signedInAuthorized;

  test('renders a button', async done => {
    const { container } = render(
      getFullyWrappedComponent(
        <NavPrimaryItem
          label={'Org Look Up'}
          icon={Frag}
          path={'/org'}
          pathname={'/org'}
          name={'orglookup'}
          closeDrawer={jest.fn()}
        />,
        { authContextProps, authObjProps }
      )
    );
    expect(container).toMatchSnapshot();
    done();
  });

  test('navigation triggers a route change', async done => {
    const { getByTestId, getByText } = render(
      getFullyWrappedComponent(
        <Router initialEntries={['/test/1']}>
          <NavPrimaryItem
            label={'Trigger'}
            icon={Frag}
            path={'/test/2'}
            pathname={'/test/2'}
            name={'trigger'}
            closeDrawer={jest.fn()}
          />
          <Route exact path={'/test/1'} component={TestOne} />
          <Route exact path={'/test/2'} component={TestTwo} />
        </Router>,
        { authContextProps, authObjProps }
      )
    );
    const button = await waitForElement(() =>
      getByTestId('navprimaryitem-trigger')
    );
    fireEvent.click(button);
    expect(getByText('test 2')).toBeInTheDocument();
    done();
  });

  test('renders a selected button when the link’s pathname matches the current URL’s pathname', async done => {
    const { container } = render(
      getFullyWrappedComponent(
        <Router initialEntries={['/test/2']}>
          <NavPrimaryItem
            label={'Trigger'}
            icon={Frag}
            path={'/test/2'}
            pathname={'/test/2'}
            name={'trigger'}
            closeDrawer={jest.fn()}
          />
          <Route exact path={'/test/1'} component={TestOne} />
          <Route exact path={'/test/2'} component={TestTwo} />
        </Router>,
        { authContextProps, authObjProps }
      )
    );
    expect(container.firstChild).toHaveClass('Mui-selected');
    done();
  });
});
