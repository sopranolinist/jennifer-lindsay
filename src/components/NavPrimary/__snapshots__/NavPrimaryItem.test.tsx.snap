// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screens - NavPrimaryItem renders a button 1`] = `
<div>
  <div
    aria-disabled="false"
    class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
    data-testid="navprimaryitem-orglookup"
    role="button"
    tabindex="0"
  >
    <div
      class="MuiListItemIcon-root"
    />
    <div
      class="MuiListItemText-root"
    >
      <span
        class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        Org Look Up
      </span>
    </div>
    <span
      class="MuiTouchRipple-root"
    />
  </div>
</div>
`;
